src/libstore/outputs-spec.hh

```diff
@@ -1,32 +1,91 @@
 #pragma once
 
+#include <optional>
+#include <set>
 #include <variant>
 
-#include "util.hh"
-
-#include "nlohmann/json_fwd.hpp"
+#include "json-impls.hh"
 
 namespace nix {
 
-typedef std::set<std::string> OutputNames;
+struct OutputNames : std::set<std::string> {
+    using std::set<std::string>::set;
+
+    // These need to be "inherited manually"
+    OutputNames(const std::set<std::string> & s)
+        : std::set<std::string>(s)
+    { }
+    OutputNames(std::set<std::string> && s)
+        : std::set<std::string>(s)
+    { }
 
-struct AllOutputs {
-    bool operator < (const AllOutputs & _) const { return false; }
+    /* This set should always be non-empty, so we delete this
+       constructor in order make creating empty ones by mistake harder.
+       */
+    OutputNames() = delete;
 };
 
-struct DefaultOutputs {
-    bool operator < (const DefaultOutputs & _) const { return false; }
+struct AllOutputs : std::monostate { };
```

Comment:
I also suggest to not subclass this standard class. Looking over the code, it does not seem to bring much advantage.

Reply:
https://en.cppreference.com/w/cpp/utility/variant/monostate It gives me all the comparison functions and other such things I need.

Reply:
It seems like that would be an appropriate use of `monostate`. It is a small class with a clear purpose, and by subclassing we make its meaning explicit.

What we're really doing here is declaring an algebraic data type, in a verbose, C++ way. Maybe that's why I don't see a problem, or may that's why there won't be a problem. Or maybe I'm misreading @tfc who isn't saying that this will be a problem per se and it isn't quite related to the comment about subclassing stl containers.